### ISSUE TITLE
Handle multi-part archives by splitting or partial processing

### ISSUE THREAD

**Issue by DominicMe** [nzbget/nzbget#472]
Occasionally I get failed PP jobs that I can extract just fine in Winrar in Windows. In those cases I get Winrar warning "cannot change attributes of ...." but it still extracts the file. This leads to a lot of wasted bandwith. Why is this warning/error cause a failure when it is non-critical?  Can this be fixed or otherwise is there a workaround?

**Comment by hugbug** [comment]
For extraction NZBGet uses external tool `unrar`. If it reports an error nzbget has to accept that.

Are you extracting over network?
Please post full processing log of such failed item (click on history item -> button Log -> button Save to file, preferably via pastebin.com

**Comment by DominicMe** [comment]
Looking close it seems the root cause in this case was that par failed. Attribute errors are not related to this issue. The nzb is multi rar set but normally nzbget handles these no problem. It seems some rar sets are corrupt only. Thinking about it it would be difficult for nzbget to process partial files since it's not content aware. Perhaps nzbget should (with an option enabled) attempt to extract files, at least multipar sets that are do not need par repair? Partial failure error might be usefull in that case as well.

**Comment by hugbug** [comment]
As you correctly mentioned nzbget isn't content aware. Since nzbget doesn't know which from failed files are important and which not it has to mark the whole download as failed. This is also required for automation, in order for alternatives to be tried.

Nzbget however doesn't delete downloaded files on failure leaving user an option to manually extract some of rar-files.

**Comment by DominicMe** [comment]
I get that but wouldn't it be better for nzbget to attempt to extract part of the download? Nzbget could possibly even attempt merge multiple nzb content into one based on the rar contents list. If not other apps could potentially use these partial files and request only the missing ones if possible. Not a simple task but at least it migt be worth detecting such multi-set archives and mark them differently so they can be manually reviewed or optionally just mark them as failed but still process them to get the parts that are not corrupt.
I feel like each rar set should be treated more like a seperate nzb, maybe some sort of a download splitting option based on rar sets could work? I don't recall seeing any multi-set archives that did not produce individually usable content so an option to split single download into multiple could work well.

**Comment by hugbug** [comment]
I don't think I can make it better than it is now: downloaded files are left on disk and user can try manual extraction and decide for himself if he can use some of undamaged files. After *KeepHistory*-days files will are deleted (if not moved away by user).

Automatic processing would require from NZBGet understanding of content. At the moment NZBGet can't even tell a difference between MKV and ISO. Making it content aware is beyond the programs purpose.

At the end there is still a possibility to process (unpack) files in an extension (post-processing) script if  you think that could work for certain kind of content. In that case the script would support certain kind of packaging and content (for example season packs consisting of mkv-files) but will not work with other types of content (for example collections of mp3-files). That is acceptable for an extension script but not OK for a general purpose program (NZBGet).

**Comment by DominicMe** [comment]
I understand what you mean, but I do wonder if it would not be a good idea to have an option (boolean) that had NZBGet detect and attempt to extract mutli-set archives seperately. In this case if some of the archive sets succeed but some do not the download would be marked as partial-fail(maybe same level as pp failure?). Another option could allow processing successful sets as normal reducing manual intervention and saving bandwith.
An easier solution to this could be to simply provide a single (boolean) option such as "split multi-set archives into seperate downloads" which would simply seperate these multi-set nzbs into seperate download jobs and process them individually as any other downloads. This would not require any content awareness, just detecting multipart archives which I believe is already done since nzbget can extract these. Seems like a very reasonable and completely optional feature, how does it seem to you?

**Comment by hugbug** [comment]
> This would not require any content awareness, just detecting multipart archives which I believe is already done since nzbget can extract these. 

NZBGet just executes `unrar x *.rar` and lets unrar do the job of detecting sets and unpacking them properly. In unrar can't process any of the sets it terminates with an error code and NZBGet knows that **something** gone wrong.

Even if NZBGet would detect archive sets and process them separately it would never get a chance to do that because:

>Looking close it seems the root cause in this case was that par failed.

When par-check fails the processing stops and no unpack is performed.

If you suggest NZBGet should detect which par-files belong to which rar-files and process each par/rar-set separately that complicates the task further.

> simply seperate these multi-set nzbs into seperate download jobs

This is a indeed valid feature request but that's not simple at all. Because of obfuscation it's not possible to detect sets until all real file names are known. The file names become known after the files are downloaded. With **DirectRename** the names are known earlier but it's even tricker because at the moment all file names are determined some files are already downloaded completely and other files are downloads-in-progress. Separating nzb into multiple nzbs at that moment is hard and error prone.

Furthermore even single episodes may consist of multiple archives: one for main video, another for subtitles and another for sample. NZBGet would need to be aware of naming conventions to group them together.

---

Wouldn't it be easier for all of us to not use season packs at all ? 😉

**Comment by DominicMe** [comment]
Ignore the original issue as that has been resolve, it's a feature suggestion at this point. Obfuscation would be a problem though it's not present most of the time so it would still work most of the time. Not a great solution though I agree. That being said though, wouldn't obfuscation only affect filenames and not extensions making it a non-issue?

Subtitles an other add on archives could be handled by detecting minimum filesize similar to how samples are handled. This option like most other options doesn't need to be perfect to be useful. There are always some cases where some things do not work well because of external factors like naming convention etc... In fact I rely on pp script that would fail to achieve it's goal if naming convention was bad, it still works 99% of the time saving a lot of manual renaming.
Speaking of subtitles and other add on archives getting split, that may be another feature that would allow discarding of certain parts like subs or samples that are seperated. This woud be like discarding samples feature but before it's even downloaded.
Regarding pars, yes pars would have to be matched to rars but normally they are names the same. In rare cases where pars are not split per archive set the download splitting can simply fail to split and continue as normal, no harm done in those cases.

Unfotunetely season packs can't be completely ignored because sonarr or for example can scnatch it with no way to disable and more broadly sometimes it may be the only available release. Multi-set archives are also not limited to season packs.

**Comment by hugbug** [comment]
> Obfuscation would be a problem though it's not present most of the time

I've just checked my 10 last downloads and 9 of them were obfuscated. Developing something that doesn't work with obfuscated posts is a waste of time.

> wouldn't obfuscation only affect filenames and not extensions making it a non-issue?

Extensions are usually obfuscated first. Some posts have files with totally random names (each file name is random).

> Multi-set archives are also not limited to season packs.

What other kinds do you mean? I'm not aware of them.

> Ignore the original issue

You missed the point. The feature request was initiated by a download failure. Multi-set nzbs which are not damaged are extracted no problem. The idea of separate set processing is to find undamaged sets and keep them. That requires not only a separate extracting but a different approach to handle par-sets. Currently all par-sets are checked at par-check stage. If any par-set check fails the processing stops, before unpack.

So basically that means a total rewrite of post-processing logic.

**Comment by DominicMe** [comment]
I don't see anywhere near as many obfuscated nzbs, its more like 1 in 25 for me but I suppose that is completely dependent on sources. Would be interested to know what you use if possible.

Files I see for obfuscated releases look like this practically always:

`newMxQHAKPi3QbsKM27iwomePdyvvvDQbfbH6BJig0EnF7A9.part01.rar`

par and sfv filenames match as well. Maybe it's due to sources I use, I am not sure.

Season packs are the most common by far but there are also non-tv nzb's packaged as one, games for example but there is definetely no rules what can be packaged as muti-set nzb. This is more for covering possible scenarios rather than it being common use case.

I said ignore original issue because it doesn't really relate to feature request anymore. I understand that if not for par failure multi-set nzb would have processed fine. Obviously I dont udenerstand the underlying logic as well as you. Perhaps having a manual (per nzb) option to split the downoad based on rar-sets after pos-processing could work? For example a multi-set nzb is marked as failed due to par failure (only some sets corrupt). In the Actions drop-down there could be an option that would split the download into multiple ones based on sets. This would allow the user to pp those newly created split downloads individually, some would fail others would pp successfuly.

**Comment by hugbug** [comment]
> I don't see anywhere near as many obfuscated nzbs, its more like 1 in 25 for me

You most likely have much more obfuscated posts than you think because you probably look at files after deobfuscation.

For any nzb in history click on an item, then on button "Log" and in search box type "renaming". If nothing is found the post wasn't obfuscated but in many cases you will see something like:

```
Renaming in-progress file de708cda68557404a0c5158ac72071e9cb90e6ef.vol001+02.par2 to 68380995A4980DD6016C6722CEF5E298.vol001+02.par2
Renaming in-progress file d92335bebec66d01ab2934743c6eef673d803be9.vol000+01.par2 to 68380995A4980DD6016C6722CEF5E298.vol000+01.par2
Renaming in-progress file 1efadb6bc86323d0dfdcedd62c2013f5.62 to Rp7ybP94ham7WF7GiJEo3q4YMFfnsYMr.part14.rar
Renaming in-progress file 1efadb6bc86323d0dfdcedd62c2013f5.61 to Rp7ybP94ham7WF7GiJEo3q4YMFfnsYMr.part37.rar
Renaming in-progress file 1efadb6bc86323d0dfdcedd62c2013f5.60 to Rp7ybP94ham7WF7GiJEo3q4YMFfnsYMr.part09.rar
```

> newMxQHAKPi3QbsKM27iwomePdyvvvDQbfbH6BJig0EnF7A9.part01.rar

That file looks like already deobfuscated. Obfuscated names don't have "part01.rar".

> Perhaps having a manual (per nzb) option to split the downoad based on rar-sets after pos-processing could work?

Any manual intervention breaks automation. When you open NZBGet webui to start the manual split  you may already have another nzb queued by Sonarr and downloaded. Not a big difference from what we have now.

----

Nowadays obfuscation goes as far that even after unpack you can end up with an obfuscated file name having random numbers and mkv-extension. Your only hint for a proper name in such case is nzb-file name. If you combine two such nzb-files into one you'll have an nzb-file which results in two video files which cannot be properly named automatically at all. If you consider this plus above examples with subtitles and samples you'll see that there is no solution for **automatic universal set splitting.**

Still it should be possible to support certain use cases, in particular the only one that actually matters - **season packs**. The solution needs to be content aware, it needs to understand what subtitles are and how samples are differ from main videos.

I would rather prefer if such solution would be developed by someone else in form of extension script.

I personally avoid season packs altogether and recommend everyone else do the same. In a rare case when you can't find alternatives **and** if season pack fails during processing **and** it still has some usable files you still can unpack them manually.

**Comment by DominicMe** [comment]
I looked at 20 nzbs and none of them have any logs with "renaming" search term even the ones ending with "-Obfuscated" so I am not sure what is going on here. I don't use many public nzb sources, any chance the nzb source does the deobfuscation before hand in some way?

Yes, I do see a lot of obfuscated filenames with correct extensions, this is why I setup 2 pp scripts to use nzbname and remove certain strings. It works 99.5% of the time for me. I am thinking I could actually make a script to query srrdb for example to get the 100% correct original release name though I think this would be easier as a sonarr script.

It wouldn't break automation at all but rather make manual processing easier. I mean if a job fails, sonarr can still do it's thing to snatch another release. In my case there is always somehing in queue so I always get a chance to review what sonarr adds before it downloads. Sometimes it may also be the only nzb or all nzb can fail making user intervention required.

In the last post I was no longer suggesting automatic set splitting. Manual option in the "Actions" tab to split a release based on sets after failure is what I was suggesting. Manually extracting can be quite labour intensive especially with some odd rar naming schemes. I am also not sure if par repair is carried out on remaining sets when one set fails. Since this is a manual option, even if it would split subs/samples/... it's not a realy issue as you can just delete those jobs before proceedign to pp those jobs as individuall nzbs.

I do agree that any automation with season packs splitting would be better done as a script if at all worth it.

I would avoid season packs but unfortunetely sonarr doesn't listen :)

**Comment by hugbug** [comment]
> Yes, I do see a lot of obfuscated filenames with correct extensions, this is why I setup 2 pp scripts to use nzbname and remove certain strings. It works 99.5% of the time for me.

There might be a misunderstanding what an obfuscated post is. With "obfuscated" I mean an nzb which after download (if not deobfuscated) results in a bunch of files which unrar cannot unpack. Such files in most cases don't have rar-extension. Whether archive name has a meaningful name (movie title) or an obscure set of characters doesn't matter for my definition of "obfuscated", as long as unrar can process these files.

Based on this definition "newMxQHAKPi3QbsKM27iwomePdyvvvDQbfbH6BJig0EnF7A9.part01.rar" is a non-obfuscated name (provided all other parts have the same archive name) whereas "1efadb6bc86323d0dfdcedd62c2013f5.61" is an obfuscated one.

Correct file names are stored within par2-files and nzbget restores correct names at renaming stage during post-processing or at download stage if DirectRename is active.

Another way to check if nzb is obfuscated is to open download details dialog and click on button "Files". If you see files ending with "partXXX.rar" then you have a non-obfuscated post. Such check works only for downloads which didn't start downloading yet; otherwise DirectRename may have already restored correct file names.

> any chance the nzb source does the deobfuscation before hand in some way?

Yes, if nzbs were created by posters (not created automatically from group headers).

**Comment by DominicMe** [comment]
Out of the 8 nzbs in queue 2 were obfuscated. It does seem that the private sites do de-obfuscate most releases as even the ones marked "-obfuscated" are not obfuscated most of the time.

So how does the manual option in the "Actions" tab to split a release based on sets after pp failure seem like to you?

**Comment by hugbug** [comment]
> So how does the manual option in the "Actions" tab to split a release based on sets after pp failure seem like to you?

It is already possible for scripts to put extra buttons into Postprocess-tab of download details dialog. When pushing the button the script is executed. The main question is now: who is going to write that script?

**Comment by DominicMe** [comment]
@hugbug I didn't know that at all, can you link to documentation here for reference?

**Comment by hugbug** [comment]
In v19 was added a possibility to put custom commands on settings page. It is documented [here](https://nzbget.net/extension-scripts#custom-commands). NZBGet is shipped with script [Email.py](https://github.com/nzbget/nzbget/blob/develop/scripts/EMail.py) which demonstrates how to do that.

Post-processing scripts can define per-nzb parameters which can then be set for each nzb individually. Such parameters are then shown on Postprocess-tab of download or history item. Parameters are defined using special config section `POST-PROCESSING PARAMETERS` of the script as described [here](https://nzbget.net/post-processing-scripts#post-processing-parameters).

By combining these two features we can put custom commands as buttons into Postrprocess-tab. The following script demonstrates that:

```python
#!/usr/bin/env python

##############################################################################
### NZBGET POST-PROCESSING SCRIPT                                          ###

# Execute demo command from postprocess-tab.
#
# This script adds custom command to postprocess-tab of download details and
# history details dialogs.
#
# NOTE: This script requires Python to be installed on your system.

##############################################################################
### OPTIONS                                                                ###

# Some option for settings page.
#SomeOption=test

##############################################################################
### POST-PROCESSING PARAMETERS                                             ###

# To do something useful click the button.
#DoSomething@Do domething useful

### NZBGET POST-PROCESSING SCRIPT                                          ###
##############################################################################

import os.sys

# Exit codes used by NZBGet
POSTPROCESS_SUCCESS=93
POSTPROCESS_ERROR=94
POSTPROCESS_NONE=95

print('[DETAIL] Script successfully started')
sys.stdout.flush()

# Check if the script is executed in custom command context
command = os.environ.get('NZBCP_COMMAND')
if command != None:
    if command == 'DoSomething':
        print('[INFO] Script is executed from postprocess-tab')
        print('[INFO] Doing something useful...')
    else:
        print('[ERROR] Invalid command ' + command)
        sys.exit(POSTPROCESS_ERROR)
else:
    print('[INFO] Script is executed as post-processing script')
    print('[INFO] Doing something useful... or just existing')

sys.exit(POSTPROCESS_SUCCESS)
```

As a result a new button is displayed on Postprocess-tab:

![screen_shot_2017-11-18_at_19_44_05](https://user-images.githubusercontent.com/3368402/32983706-3c138acc-cc99-11e7-847f-d3cc52cb19be.png)

That ability to add custom commands into Postprocess-tab is merely a side effect of implementing custom commands for settings page. When I discovered that side effect I was thinking if such per-nzb custom commands can be useful for anything. The per-nzb post-processing parameters exist for ever, yet I don't know any script using that feature. For that reason I have not implemented the per-nzb custom commands till the end. As a result the button displayed on the screenshot doesn't really work. I forgot about this when I was writing my previous post.

If you think you can write a script which will do the processing of damaged multiset archives and therefore you could use the per-nzb custom commands I could implement them completely.

Or maybe it would be even better to implement an ability to put custom commands into Actions menu instead.
